musig/docs/musig-spec.md

```diff
@@ -0,0 +1,182 @@
+    T.challenge_scalar(label) -> scalar
+    T.challenge_scalar(label) == T.challenge_bytes<64>(label) mod |G|
+    ```
+
+
+## Single-message signature
+
+Single-message signature is a Schnorr proof of knowledge of a secret [scalar](#scalar)
+corresponding to some [verification key](#verification-key) in a context of some _message_.
+
+The protocol is the following:
+
+1. Prover and verifier obtain a [transcript](#transcript) `T` that is assumed to be already bound to the _message_ being signed.
+2. Prover and verifier both commit the verification key `X` (computed by the prover as `X = x·B`):
+    ```
+    T.commit("X", X)
+    ```
+3. Prover creates a _secret nonce_: a randomly sampled [scalar](#scalar) `r`.
+4. Prover commits to its nonce:
+    ```
+    R = r·B
+    ```
+5. Prover sends `R` to the verifier.
+6. Prover and verifier write the nonce commitment `R` to the transcript:
+    ```
+    T.commit("R", R)
+    ```
+7. Prover and verifier compute a Fiat-Shamir challenge scalar `e` using the transcript:
+    ```
+    c = T.challenge_scalar("c")
+    ```
+8. Prover blinds the secret scalar `x` using the nonce and the challenge:
+    ```
+    s = r + c·x
+    ```
+9. Prover sends `s` to the verifier.
+10. Verifier checks the relation:
+    ```
+    s·B  ==  R + c·X
+    ```
+
+
+## Multi-message signature
+
+Multi-message signature is a Schnorr proof of knowledge of a set of secret [scalars](#scalar)
+corresponding to a set of [verification keys](#verification-key) in a context of some _message_,
+and a list of per-key _submessages_. The goal of the protocol is to safely permit the signers to verify
+only their individual submessages, ignoring other signers’ submessages.
+
+1. Prover and verifier obtain a [transcript](#transcript) `T` that is assumed to be already bound to the _message_ being signed.
+2. Prover and verifier both commit the set of `n` verification keys `X[i]` and submessages `m[i]`:
+    ```
+    T.commit_bytes("Musig.Multimessage", LE64(n))
+    T.commit_bytes("X", X[0])
+    T.commit_bytes("m", m[0])
+    ...
+    T.commit_bytes("X", X[n-1])
+    T.commit_bytes("m", m[n-1])
+    ```
+3. Prover creates a _secret nonce_: a randomly sampled [scalar](#scalar) `r`.
+4. Prover commits to its nonce:
+    ```
+    R = r·B
+    ```
+5. Prover sends `R` to the verifier.
+6. Prover and verifier write the nonce commitment `R` to the transcript:
+    ```
+    T.commit("R", R)
+    ```
+7. Prover and verifier compute a Fiat-Shamir challenge scalar `e[i]` for each `i`th key, using the _copy_ of transcript `T`:
+    ```
+    T’ = copy(T)
+    T’.commit_bytes("i", LE64(i))
+    c[i] = T’.challenge_scalar("c")
+    ```
+8. Prover blinds the secret scalars `x[i]` using the nonce and the challenges `c[i]`:
+    ```
+    s = r + sum{c[i]·x[i]}
+    ```
+9. Prover sends `s` to the verifier.
+10. Verifier checks the relation:
+    ```
+    s·B  ==  R + sum{c[i]·X[i]}
+    ```
+
+## Key aggregation
+
+TBD: spec for offline key aggregation.
```

Comment:
Link to the key aggregation notes?
https://github.com/interstellar/slingshot/tree/7c9e5474e133720492261f7b8171c075b0de41dd/musig#key-aggregation